File: 2012-sum-of-beauty-in-the-array/2012-sum-of-beauty-in-the-array.java
class Solution {
    public int sumOfBeauties(int[] nums) {
        int n = nums.length;
        int ans = 0;
        int[] prefix_max = new int[n];
        int[] suffix_min = new int[n];
        prefix_max[0] = nums[0];
        suffix_min[n-1] = nums[n-1];
        
        for(int i=1; i<n; i++) {
            prefix_max[i] = Math.max(nums[i-1], prefix_max[i-1]);
        }
        for(int i=n-2; i>=0; i--) {
            suffix_min[i] = Math.min(nums[i+1], suffix_min[i+1]);
        }
        
        
        for(int i=1; i<n-1; i++) {
            if(prefix_max[i] < nums[i] && nums[i] < suffix_min[i]) {
                  ans += 2;
            }
            else if(nums[i-1] < nums[i] && nums[i] < nums[i+1]) {
                 ans += 1;
            }
        }
        return ans;       
            
        
    }
}

File: 2012-sum-of-beauty-in-the-array/2012-sum-of-beauty-in-the-array.py
class Solution:
    def sumOfBeauties(self, nums: List[int]) -> int:
        prefix = [-1 for _ in range(len(nums))]
        suffix = [-1 for _ in range(len(nums))]
        msf = -1 # Max so far
        prefix[0] = nums[0]
        suffix[-1] = nums[-1]
        for i in range(1, len(nums)):
            prefix[i] = max(nums[i-1], prefix[i-1])
        
        msf = -1 # Max so far reinitialized
        for i in range(len(nums)-2, -1, -1):
            suffix[i] = min(nums[i+1], suffix[i+1])
        
        # print(prefix)
        # print(suffix)
        flag = 0
        ans = 0
        for i in range(1, len(nums)-1):
            if prefix[i] < nums[i] < suffix[i]:
                ans += 2
            elif nums[i-1] < nums[i] < nums[i+1]:
                ans += 1
        return ans
        
        